DEVX-2467: Create consumer lag client use case

awalther28 opened the pull request:
### Description 

https://confluentinc.atlassian.net/browse/DEVX-2467
_What behavior does this PR change, and why?_
Adds:
 - consumer scenario section
 - consumer lag scenario
 - kafka-lag-exporter with automated and step by step setup
 - update docker to use scale functionality
 - clean up consumer dashboard
 - update prometheus scrape configs per docker container name changes

Additional things:
- fix producer key so that records are produced to all topics


Docs PR: https://github.com/confluentinc/docs-platform/pull/462
Staging link: http://staging-docs-independent.confluent.io.s3-website-us-west-2.amazonaws.com/docs-platform/PR/462/6.1.0/tutorials/examples/ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.html 


### Author Validation

_Describe the validation already done, or needs to be done, by the PR submitter._

<!-- Uncomment any of the following that are required -->
- [x] Documentation
- [x] ccloud-monitoring
<!-- - [ ] ccloud/beginner-cloud -->
<!-- - [ ] ccloud/ccloud-stack -->
<!-- - [ ] clickstream -->
<!-- - [ ] clients/avro -->
<!-- - [ ] clients/cloud -->
<!-- - [ ] cloud-etl -->
<!-- - [ ] connect-streams-pipeline -->
<!-- - [ ] cp-quickstart -->
<!-- - [ ] kubernetes/gke-base -->
<!-- - [ ] kubernetes/replicator-gke-cc -->
<!-- - [ ] microservices-orders -->
<!-- - [ ] multi-datacenter -->
<!-- - [ ] multiregion -->
<!-- - [ ] replicator-schema-translation -->
<!-- - [ ] replicator-security -->
<!-- - [ ] security/rbac -->
<!-- - [ ] security/secret-protection -->


### Reviewer Tasks

_Describe the tasks/validation that the PR submitter is requesting to be done by the reviewer._

<!-- Uncomment any of the following that are required -->
- [] Documentation
- [] ccloud-monitoring
<!-- - [ ] Documentation -->
<!-- - [ ] ccloud/beginner-cloud -->
<!-- - [ ] ccloud/ccloud-stack -->
<!-- - [ ] clickstream -->
<!-- - [ ] clients/avro -->
<!-- - [ ] clients/cloud -->
<!-- - [ ] cloud-etl -->
<!-- - [ ] connect-streams-pipeline -->
<!-- - [ ] cp-quickstart -->
<!-- - [ ] kubernetes/gke-base -->
<!-- - [ ] kubernetes/replicator-gke-cc -->
<!-- - [ ] microservices-orders -->
<!-- - [ ] multi-datacenter -->
<!-- - [ ] multiregion -->
<!-- - [ ] replicator-schema-translation -->
<!-- - [ ] replicator-security -->
<!-- - [ ] security/rbac -->
<!-- - [ ] security/secret-protection -->

awalther28 committed:
DEVX-2467: add base kafka-lag components

awalther28 committed:
DEVX-2467: fix kafka-lag-exporter app conf

awalther28 committed:
DEVX-2467: clean up consumer dashboard

awalther28 committed:
DEVX-2467: change docker, add images

awalther28 committed:
DEVX-2466: update docs, fix partitioning, add threshold

awalther28 committed:
DEVX-2467: update with new info

awalther28 committed:
DEVX-2467: add picture

awalther28 committed:
DEVX-2467: breakout pictures of cg dashboard

awalther28 committed:
DEVX-2467: wrap up consumer lag docs

awalther28 committed:
DEVX-2467: add 2 other steps

awalther28 committed:
DEVX-2467: fix indentation

awalther28 committed:
DEVX-2467: add mbean info to scenario

awalther28 committed:
DEVX-2467: clean up intro

awalther28 committed:
Merge branch '6.1.0-post' into DEVX-2467

ybyzek commented on ccloud-monitoring/docker-compose.yml:
```diff
@@ -5,9 +5,11 @@ volumes:
   cap_add:
     - NET_ADMIN
-x-consumer: &consumer
+x-client-base: &client-base
```
Very cool docker-compose super powers @awalther28

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+Consumer Lag Problem
+********************
+
+Consumer group lag is a tremendous performance indicator. It tells you the difference between the
```
```suggestion
Consumer lag is a tremendous performance indicator. It tells you the offset difference between the
```

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+consumer groups or concepts like committing offsets, please refer to this
+`Kafka Consumer documentation <https://docs.confluent.io/platform/current/clients/consumer.html>`__.
+
+A large or quickly growing lag indicates that the consumer is not able to keep up with
```
```suggestion
A large consumer lag, or a quickly growing lag, indicates that the consumer is not able to keep up with
```

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+A large or quickly growing lag indicates that the consumer is not able to keep up with
+the volume of messages on a topic.
+
+This scenario will look at Confluent Cloud metrics from the Metrics API, `kafka-lag-exporter <https://github.com/lightbend/kafka-lag-exporter>`__ metrics, and
```
Please use `|ccloud|` instead of `Confluent Cloud`, here and wherever else is applicable

awalther28 replied:
Idk why that is such a hard one for me to remember 😓  will correct throughout these additions

awalther28 replied:
corrected in latest commit f3f5b8f

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+Introduce failure scenario
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+#. Stop ``consumer-1`` container, thus removing a consumer from the consumer group and add 2 producers:
```
- Presumably this command stops ``consumer-1`` because the scale is dropped from 2 to 1, but maybe it's worth an explanation.  I had to do a double-take on this to understand that.
- Add `2` producers?
- Maybe it's worth a reminder to say, something like, by default, X producer and Y consumers are running.  Adjust that to X' and Y' in order to show...

awalther28 replied:
👍

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+
+   |Consumer Rebalance Bump|
+
+   - An upward trend in ``Consumer group lag in records``.  ``Consumer group lag in seconds`` will have a less dramatic increase. Both indicating that the producer is creating more messages than the consumer can fetch in a timely manner. These metrics are derived from ``kafka-lag-exporter``.
```
If ``kafka-lag-exporter`` is not an MBean, explain what is providing those stats (e.g. explain the docker container, what it does, how it works, etc -- not verbose, just a sentence or two)

awalther28 replied:
👍

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+      demo-cloud-monitoring-1 demo-topic-1    1          48097           48151           54              consumer-demo-cloud-monitoring-1-1-b0bec0b5-ec84-4233-9d3e-09d132b9a3c7 /10.2.10.251    consumer-demo-cloud-monitoring-1-1
+      demo-cloud-monitoring-1 demo-topic-1    2          48310           48370           60              consumer-demo-cloud-monitoring-1-1-b0bec0b5-ec84-4233-9d3e-09d132b9a3c7 /10.2.10.251    consumer-demo-cloud-monitoring-1-1
+
+   The downside of this view is the lack of historical context that the ``Consumer Client Metrics`` dashboard provides.
```
```suggestion
   This command provides a snapshot in time, but it lacks of historical context that the ``Consumer Client Metrics`` dashboard provides.
```

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+
+   The downside of this view is the lack of historical context that the ``Consumer Client Metrics`` dashboard provides.
+
+#. Another view of consumer lag can be found in |ccloud|. Open the UI, navigate to the "Consumers" section and click on the ``demo-cloud-monitoring-1`` consumer group.
```
Maybe CCloud UI should precede the command line? (swap this with above)

awalther28 replied:
👍

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+
+   |Confluent Cloud Request Increase|
+
+#. The consumer logs won't show that the consumer is falling behind which is why it is important to keep an eye on consumer group metrics.
```
```suggestion
#. The consumer logs won't show that the consumer is falling behind which is why it is important to have a robust monitoring solution that covers consumer lag.
```

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst:
```diff
@@ -0,0 +1,164 @@
+   |Confluent Cloud Request Increase|
+
+#. The consumer logs won't show that the consumer is falling behind which is why it is important to keep an eye on consumer group metrics.
+   Logs will tell us that the consumer group rebalanced, the logs can be accessed with the following command:
```
```suggestion
   However, consumer logs will show that the consumer group rebalanced.  The logs can be accessed with the following command:
```

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/overview.rst:
```diff
@@ -0,0 +1,19 @@
+other non-java clients--they generally offer similar metrics.
+
+The source code for the client can found in the
+devx-examples:`examples/ccloud-monitoring/src|examples/ccloud-monitoring/src` directory. The client
```
```suggestion
devx-examples:`ccloud-monitoring/src|ccloud-monitoring/src` directory. The client
```

@awalther28 I messed this up originally, but it's being fixed in the other scenarios in my other open PR (update -- which I had recently merged)

ybyzek commented on ccloud-monitoring/start.sh:
```diff
@@ -48,6 +48,9 @@ eval $CMD \
     && print_code_pass -c "source $DELTA_CONFIGS_ENV" \
     || exit_with_error -c $? -n "$NAME" -m "$CMD" -l $(($LINENO -3))
 
+printf "\n====== Allow service account to describe Confluent Cloud cluster and consumer groups\n"
```
Thanks for identifying this ACL gap.  @awalther28 @rspurgeon my leaning is to move this into `ccloud-stack` creation (`utils/ccloud_library.sh`) for all other future use cases to benefit.  Thoughts?

awalther28 replied:
I think that would be ideal :) happy to move it into there in this PR.

ybyzek replied:
Yeah, let's do it.

awalther28 replied:
Moved, updated the start script and instructions. I validated the start script and tearing the cluster down. 
A few unrelated ACLs weren't being deleted in utils/ccloud_library.sh, so I added them to be deleted as well.

ybyzek reviewed (commented):
@awalther28 first pass complete!  I haven't run it yet...will do so soon

awalther28 committed:
Apply suggestions from code review

Co-authored-by: Yeva Byzek <ybyzek@users.noreply.github.com>

awalther28 committed:
DEVX-2467: change file name, apply yeva suggestions

ybyzek commented on ccloud-monitoring/docs/index.rst:
```diff
@@ -9,6 +9,7 @@ Monitoring |ak-tm| Clients to Confluent Cloud
 
    Monitoring overview and setup <monitoring-overview>
    Producer scenarios <producer-scenarios>
+   Consumer scenarios <consumer-scenarios>
```
Sorry for slipping this comment into this PR but WDYT about moving Additional resources to be after Clean up?

awalther28 replied:
Moved. Rerunning docs build rn to validate changes didn't break anything.

awalther28 committed:
DEVX-2467: mv acl creation, update doc layout and instructions

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+Introduce failure scenario
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+#. By default 2 consumers and 1 producer are running. Adjust that to 1 consumer and 5 producers in order to show an increase in consumer lag.
```
Just curious, why is the default 2 consumers?  Why isn't it 1 consumer?  If each consumer-related scenario can scale the number of consumers as it needs to, would it make sense (for footprint and ease of explaining) to make the default 1 consumer?  (Plus, the unique dashboard view of consumer rebalances could be moved to a new scenario)

awalther28 replied:
Great point. Now that the docker-compose has clients in state to scale with an easy command, I should probably remove the extra consumer. 
Moments like these are why I love PRs.

awalther28 replied:
:sonic_waiting: I need to retake screenshots of the scenario, actively working on it.

awalther28 replied:
updated the screenshots in https://github.com/confluentinc/examples/pull/948/commits/9b6e040d6ad2773e51a2e0a49e50b944549ac902

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+Introduce failure scenario
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+#. By default 2 consumers and 1 producer are running. Adjust that to 1 consumer and 5 producers in order to show an increase in consumer lag.
```
```suggestion
#. By default 2 consumers and 1 producer are running. Adjust that to 1 consumer and 5 producers in order to force the condition where the consumer cannot keep up with the rate of messages being produced, which will cause an increase in consumer lag.
```

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+   |Consumer Rebalance Bump|
+
+   - An upward trend in ``Consumer group lag in records``.  ``Consumer group lag in seconds`` will have a less dramatic increase.
+     Both indicating that the producer is creating more messages than the consumer can fetch in a timely manner.
```
```suggestion
     Both indicate that the producer is creating more messages than the consumer can fetch in a timely manner.
```

ybyzek reviewed (commented):
All producers have the same client id:

```
producer_1            | 	client.id = producer-1
producer_2            | 	client.id = producer-1
producer_3            | 	client.id = producer-1
producer_4            | 	client.id = producer-1
producer_5            | 	client.id = producer-1
```

In a real world scenario, they would be configured with a unique client id per producer.  Advantages: they show up uniquely in Grafana & CCloud & logs, etc.  (whereas right now it looks like all one producer, except from docker container perspective).  Do want to fix this?  If yes, it will make the scale up code a bit trickier.  Ideas:

1. Find a way with Docker Compose. I'm not familiar with `scale` options but I think I found a way a long time ago...without `scale`...to set an env parameter, e.g. `ID=<value>`, and calling the producer 4 more times and incrementing ID each time. (I can explain more in Slack if you want to go this route)
2. Assign a random unique `client.id` every time the producer app starts up (do it from within the Java app code itself).  This would enable you to still use `scale` (which is pretty slick)

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+
+   |Consumer Fetch Increase|
+
+   - All of the graphs in the ``Throughput`` are indicating the consumer is processing more bytes/records.
```
For this point and the previous point, while they are true, it doesn't necessarily mean that the consumer lag would necessarily increase (especially if the consumer is tuned/resources properly).  Would it make sense to make a note of this?

awalther28 replied:
I added a note. 
![Screen Shot 2021-02-25 at 10 58 35 AM](https://user-images.githubusercontent.com/12937975/109196992-ab9e3900-7759-11eb-89fe-abc10ad489ae.png)

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+
+   |Consumer Throughput Increase|
+
+#. Another view of consumer lag can be found in |ccloud|. Open the UI, navigate to the "Consumers" section and click on the ``demo-cloud-monitoring-1`` consumer group.
```
Great tie in!

awalther28 commented:
@ybyzek to your note "In a real world scenario, they would be configured with a unique client id per producer." Right now we don't monitor the extra producers or consumers, their metrics aren't scraped by prometheus. It would be simple to instruct someone to do it (add the "container name:1234" to a targets list in monitoring_configs/prometheus/prometheus.yml and restart prometheus). 
While I want to fix this, my leaning is to wait to fix it until we need to look at the scaled producer/consumer logs or metrics in a scenario. (Aka when consumer group rebalances are discussed.) If you prefer it's fixed now, I can certainly work that in asap.

ybyzek commented:
> While I want to fix this, my leaning is to wait to fix it until we need to look at the scaled producer/consumer logs or metrics in a scenario. (Aka when consumer group rebalances are discussed.) If you prefer it's fixed now, I can certainly work that in asap.

It's your call.  I don't think current implementation detracts from the scenario, so it's ok to bump it into a separate issue to track for later prioritization.

Thanks for explaining that the extra clients aren't scraped by Prometheus, so I guess it won't show up as a "gap" in Grafana.  However, one place it does still show up is CCloud UI...like in the Dataflow it shows only 1 producer instead of 5.  This is not a blocker in any way, just wanted to mention it.

awalther28 committed:
Apply suggestions from code review

Co-authored-by: Yeva Byzek <ybyzek@users.noreply.github.com>

awalther28 committed:
DEVX-2467: start of removal of extra consumer

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+   - An upward trend in ``Consumer group lag in records``.  ``Consumer group lag in seconds`` will have a less dramatic increase.
+     Both indicate that the producer is creating more messages than the consumer can fetch in a timely manner.
+     These metrics are derived from the ``kafka-lag-exporter`` container.
+     `kafka-lag-exporter <https://github.com/lightbend/kafka-lag-exporter>`__ is a scala open source project that collects data about consumer groups and presents them in a Prometheus scrapable format.
```
Great explanation!  For consideration, move to the intro?

awalther28 replied:
Are you sure? Earlier it seemed like you wanted an explanation of this sort here (https://github.com/confluentinc/examples/pull/948#discussion_r582325128).

ybyzek replied:
My leaning:

- In the introduction it could introduce the Kafka Lag Exporter overall (with the longer explanation to be able to answer what is Kafka Lag Exporter
- In this step it could say that consumer lag is derived from ``kafka-lag-exporter``.  User by this point should already know what Kafka Lag Exporter is, you are just reminding them that this metric. is derived from that

awalther28 replied:
I moved the kafka-lag-exporter info to the top in commit https://github.com/confluentinc/examples/pull/948/commits/9b6e040d6ad2773e51a2e0a49e50b944549ac902

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+
+   .. code-block:: bash
+
+      docker-compose up -d --scale consumer=1 --scale producer=5
```
If running this without setting the `EXPORT` commands printed at the end of the start script, the output is:

```
[8:08:37] ~/git/examples/ccloud-monitoring(DEVX-2467) ✗: docker-compose up -d --scale consumer=1 --scale producer=5
WARNING: The METRICS_API_KEY variable is not set. Defaulting to a blank string.
WARNING: The METRICS_API_SECRET variable is not set. Defaulting to a blank string.
WARNING: The CLOUD_CLUSTER variable is not set. Defaulting to a blank string.
WARNING: The CONFIG_FILE variable is not set. Defaulting to a blank string.
WARNING: The BOOTSTRAP_SERVERS variable is not set. Defaulting to a blank string.
WARNING: The SASL_JAAS_CONFIG variable is not set. Defaulting to a blank string.
Recreating kafka-lag-exporter ... 
Recreating ccloud-exporter    ... 
node-exporter is up-to-date
Recreating kafka-lag-exporter                      ... done
Recreating ccloud-exporter                         ... done
Stopping and removing ccloud-monitoring_consumer_2 ... done
Recreating ccloud-monitoring_producer_1            ... done
Recreating ccloud-monitoring_consumer_1            ... done
Creating ccloud-monitoring_producer_2              ... done
Creating ccloud-monitoring_producer_3              ... done
Creating ccloud-monitoring_producer_4              ... done
Creating ccloud-monitoring_producer_5              ... done
```

- So should those export statements be executed by the user?  (if yes, that was one advantage of adding those to a `.env` file)
- Should the clients not have dependencies on the monitoring docker containers?  (so they don't get updated when scaling up/down)

awalther28 replied:
A .env file would resolve the problem when rebringing up all the containers during docker-compoes --scale but doesn't resolve why all of the containers are being evaluated to be brought up. 
Currently the only dependencies are the clients on prometheus + grafana (u right, this ain't needed). Given that, it doesn't make sense that kafka-lag-exporter, node-exporter, ccloud-exporter are recreated here. I'm digging into this.

awalther28 replied:
I removed the client dependencies in commit https://github.com/confluentinc/examples/pull/948/commits/9b6e040d6ad2773e51a2e0a49e50b944549ac902 but still see all of the containers being evaluated for "updates" when scaling. A brief look into `docker-compose up --scale` this morning didn't yield any obvious ways to circumvent this. I added an "export" of the necessary env vars to a `.env` file so that these variables are set when a --scale command is run.

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+
+   |Confluent Cloud Consumer Lag|
+
+   This provides a snapshot in time, but it lacks of historical context that the ``Consumer Client Metrics`` dashboard provides.
```
```suggestion
   This provides a snapshot in time, but it lacks the historical view that the ``Consumer Client Metrics`` dashboard provides.
```

awalther28 replied:
resolved in commit https://github.com/confluentinc/examples/pull/948/commits/9b6e040d6ad2773e51a2e0a49e50b944549ac902

ybyzek commented on ccloud-monitoring/docs/consumer-scenarios/increasing-consumer-lag.rst:
```diff
@@ -0,0 +1,168 @@
+This scenario will look at |ccloud| metrics from the Metrics API, `kafka-lag-exporter <https://github.com/lightbend/kafka-lag-exporter>`__ metrics, and
+client metrics from the client application’s MBean object ``kafka.consumer:type=consumer-fetch-manager-metrics,client-id=<client_id>``.
+
+Introduce failure scenario
```
It's not really a *failure* scenario, it might be misleading to call it that.

awalther28 replied:
I would argue that teams do have SLAs to process the latest messages within a period of time, so having a lag (if big enough) would could cause you to break that SLA. Imagine going from 25 second delays to 12 minute delays (possible if consumer code is tweaked), that could be a failure to meet your SLA of processing messages within a 30 second delay. 
It's definitely true that this lag introduced here isn't that massive. I can add a back story in the intro about how this could turn into a failure scenario but with resource restrictions, we will show a "degraded state" (barely). Or change the title to "Introduce degraded state scenario" (imo this is still misleading given the small lag that is introduced).

awalther28 commented:
For time's sake, I'm going to bump the produce/consumer client id issue to some later date. Created this jira to track it: https://confluentinc.atlassian.net/browse/DEVX-2531. I did spend some time yesterday evening trying to figure out how to get the scale # into the container, it's not straightforward. We could use unique container ID's as the client IDs 🤷

awalther28 committed:
DEVX-2467: update screenshots

awalther28 committed:
Merge branch 'DEVX-2467' of github.com:confluentinc/examples into DEVX-2467

awalther28 committed:
DEVX-2467: simplify start instructions

awalther28 committed:
DEVX-2467: fix docs link

awalther28 committed:
DEVX-2467: fix grammar

awalther28 commented:
@ybyzek I have addressed all of your comments, either by making updates or asking further questions. The staging docs should reflect the most up to date content.

ybyzek reviewed (approved):
LGTM @awalther28, very well done!
